packages/babel-core/src/transformation/util/clone-deep.ts

```diff
@@ -1,33 +1,66 @@
-//https://github.com/babel/babel/pull/14583#discussion_r882828856
-function deepClone(value: any, cache: Map<any, any>): any {
+const circleSet = new Set();
+// https://github.com/babel/babel/pull/14583#discussion_r882828856
+function deepClone(
+  value: any,
+  cache: Map<any, any>,
+  allowCircle: boolean,
+): any {
   if (value !== null) {
-    if (cache.has(value)) return cache.get(value);
+    if (allowCircle) {
+      if (cache.has(value)) return cache.get(value);
+    } else {
+      if (circleSet.has(value)) {
+        circleSet.clear();
+        throw new Error("Babel-deepClone: Cycles are not allowed in AST");
+      }
+      circleSet.add(value);
```

Comment:
How big is the performance impact of using this set compared to:
- the old behavior with the map
- the behavior with no protection against infinite loops

If it's significantly expensive, we could keep track of the depth we are at in the cloning and only start using the cycle detection logic if it's, for example, >100. It means that we'll be slower in error cases, but faster in the (common) successful case.

Reply:
current: 263 ops/sec ±1.13% (3.809ms)
circle-check: 255 ops/sec ±0.9% (3.924ms)
no-circle-check: 357 ops/sec ±1.55% (2.799ms)
circle-check-100-depth: 355 ops/sec ±0.7% (2.814ms)

I'll add depth detection. :)